### PR TITLE
feat(core): add workspace billing

### PR DESCRIPTION
**Pull request by JimmFly**

### CODE REVIEW

**Comment by JimmFly** [comment]
* **#9043** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9043?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9043?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8920** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8920?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#9026](https://github.com/toeverything/AFFiNE/pull/9026) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9026?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#9085](https://github.com/toeverything/AFFiNE/pull/9085) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9085?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `canary`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx [hunk 1]
```diff
@@ -41,8 +41,10 @@ import * as styles from './styles.css';
 export const WorkspaceSettingBilling = () => {
   const t = useI18n();
   const workspace = useService(WorkspaceService).workspace;
-  const subscriptionService = useService(SubscriptionService);
-  const team = useLiveData(subscriptionService.subscription.team$);
+  const subscriptionService = useService(WorkspaceSubscriptionService);
+  const subscription = useLiveData(
+    subscriptionService.subscription.subscription$
```
Comment (EYHN) [inline packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx:46]:
```suggestion
    subscriptionService.subscription.team$
```

File: packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx [hunk 2]
```diff
@@ -81,8 +83,10 @@ export const WorkspaceSettingBilling = () => {
 
 const TeamCard = () => {
   const t = useI18n();
-  const subscriptionService = useService(SubscriptionService);
-  const teamSubscription = useLiveData(subscriptionService.subscription.team$);
+  const subscriptionService = useService(WorkspaceSubscriptionService);
+  const teamSubscription = useLiveData(
+    subscriptionService.subscription.subscription$
```
Comment (EYHN) [inline packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx:88]:
```suggestion
    subscriptionService.subscription.team$
```

File: packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx [hunk 3]
```diff
@@ -198,23 +202,20 @@ const ResumeSubscription = ({ expirationDate }: { expirationDate: string }) => {
 
 const TypeFormLink = () => {
   const t = useI18n();
-  const subscriptionService = useService(SubscriptionService);
+  const subscriptionService = useService(WorkspaceSubscriptionService);
   const authService = useService(AuthService);
 
-  const team = useLiveData(subscriptionService.subscription.team$);
+  const team = useLiveData(subscriptionService.subscription.subscription$);
```
Comment (EYHN) [inline packages/frontend/core/src/desktop/dialogs/setting/workspace-setting/billing/index.tsx]:
```suggestion
  const team = useLiveData(subscriptionService.subscription.team$);
```

File: packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    exhaustMapSwitchUntilChanged(
+      (a, b) => a.currentWorkspaceId === b.currentWorkspaceId,
+      ({ currentWorkspaceId }) => {
+        return fromPromise(async signal => {
+          if (!currentWorkspaceId) {
+            return undefined; // no subscription if no user
+          }
+
+          const serverConfig =
+            await this.serverService.server.features$.waitForNonNull(signal);
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts]:
```suggestion
  revalidate = effect(
    exhaustMapWithTrailing(
      () => {
        return fromPromise(async signal => {
          const currentWorkspaceId = this.workspaceService.workspace.id
          if (!currentWorkspaceId) {
            return undefined; // no subscription if no user
          }

          const serverConfig =
            await this.serverService.server.features$.waitForNonNull(signal);
```

File: packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+          }
+          const { workspaceId, subscription } =
+            await this.store.fetchWorkspaceSubscriptions(
+              currentWorkspaceId,
+              signal
+            );
+          if (currentWorkspaceId !== workspaceId) {
+            // The user has changed, ignore the result
+            return null;
+          }
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts]:
```suggestion
          
```
never happened

File: packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        if (!currentWorkspaceId) {
+          this.subscription$.next(null);
+        } else {
+          this.subscription$.next(
+            this.store.getCachedWorkspaceSubscription(currentWorkspaceId)
+          );
+        }
+      }
+    )
+  );
```
Comment (EYHN) [inline packages/frontend/core/src/modules/cloud/entities/workspace-subscription.ts]:
```suggestion
        );
      }
    )
  );
```

File: packages/frontend/graphql/src/graphql/workspace-invoices-count.gql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+query getWorkspaceInvoicesCount($workspaceId: String!) {
+  workspace(id: $workspaceId) {
+    invoiceCount
```
Comment (forehalo) [inline packages/frontend/graphql/src/graphql/workspace-invoices-count.gql]: use workspaceInvoices ?

Comment (JimmFly) [reply packages/frontend/graphql/src/graphql/workspace-invoices-count.gql]: It looks like we don't really need this `getWorkspaceInvoicesCountQuery`, I will delete it.

**Review by forehalo — APPROVED** [review]

**Pull request merged**